internal/dump/archive.go

```diff
@@ -75,58 +89,93 @@ const (
 	c_ISVTX = 01000 // Save text (sticky bit)
 )
 
-func tarNode(ctx context.Context, tw *tar.Writer, node *restic.Node, repo restic.Repository) error {
+func archiveNode(ctx context.Context, writer interface{}, node *restic.Node, repo restic.Repository) error {
+	var w io.Writer
 	relPath, err := filepath.Rel("/", node.Path)
 	if err != nil {
 		return err
 	}
 
-	header := &tar.Header{
-		Name:       filepath.ToSlash(relPath),
-		Size:       int64(node.Size),
-		Mode:       int64(node.Mode.Perm()), // c_IS* constants are added later
-		Uid:        int(node.UID),
-		Gid:        int(node.GID),
-		Uname:      node.User,
-		Gname:      node.Group,
-		ModTime:    node.ModTime,
-		AccessTime: node.AccessTime,
-		ChangeTime: node.ChangeTime,
-		PAXRecords: parseXattrs(node.ExtendedAttributes),
-	}
+	switch writer.(type) {
+	case *tar.Writer:
+		tw := writer.(*tar.Writer)
+		header := &tar.Header{
+			Name:       filepath.ToSlash(relPath),
+			Size:       int64(node.Size),
+			Mode:       int64(node.Mode.Perm()), // c_IS* constants are added later
+			Uid:        int(node.UID),
+			Gid:        int(node.GID),
+			Uname:      node.User,
+			Gname:      node.Group,
+			ModTime:    node.ModTime,
+			AccessTime: node.AccessTime,
+			ChangeTime: node.ChangeTime,
+			PAXRecords: parseXattrs(node.ExtendedAttributes),
+		}
 
-	// adapted from archive/tar.FileInfoHeader
-	if node.Mode&os.ModeSetuid != 0 {
-		header.Mode |= c_ISUID
-	}
-	if node.Mode&os.ModeSetgid != 0 {
-		header.Mode |= c_ISGID
-	}
-	if node.Mode&os.ModeSticky != 0 {
-		header.Mode |= c_ISVTX
-	}
+		// adapted from archive/tar.FileInfoHeader
+		if node.Mode&os.ModeSetuid != 0 {
+			header.Mode |= c_ISUID
+		}
+		if node.Mode&os.ModeSetgid != 0 {
+			header.Mode |= c_ISGID
+		}
+		if node.Mode&os.ModeSticky != 0 {
+			header.Mode |= c_ISVTX
+		}
 
-	if IsFile(node) {
-		header.Typeflag = tar.TypeReg
-	}
+		if IsFile(node) {
+			header.Typeflag = tar.TypeReg
+		}
 
-	if IsLink(node) {
-		header.Typeflag = tar.TypeSymlink
-		header.Linkname = node.LinkTarget
-	}
+		if IsLink(node) {
+			header.Typeflag = tar.TypeSymlink
+			header.Linkname = node.LinkTarget
+		}
 
-	if IsDir(node) {
-		header.Typeflag = tar.TypeDir
-		header.Name += "/"
-	}
+		if IsDir(node) {
+			header.Typeflag = tar.TypeDir
+			header.Name += "/"
+		}
 
-	err = tw.WriteHeader(header)
+		err = tw.WriteHeader(header)
+
+		if err != nil {
+			return errors.Wrap(err, "TarHeader ")
+		}
+
+		w = tw
+	case *zip.Writer:
+		header := &zip.FileHeader{
+			Name:               strings.TrimPrefix(node.Path, "/"),
+			UncompressedSize64: uint64(node.Size),
+			Modified:           node.ModTime,
+		}
+		if IsLink(node) {
+			header.SetMode(0755 | os.ModeSymlink)
+		}
```

Comment:
If you just call `header.SetMode(node.Mode)` unconditionally it will keep read-only attribute (the lower byte of `ExternalAttrs`) on Windows.
And as far as I know, ZIP supports symlinks only by special extension "UNIX Extra Field (0x000d)".

Reply:
Fixed. Symlinks seem to work in both zip and tar with or without the change but `node.Mode` seems to be a better thing to use. Thanks!